rust/blockstore/src/arrow/block/types.rs

```diff
@@ -158,36 +169,103 @@ impl Block {
-                // But in our case NaN should not be allowed so we should always have total order
-                Some(k) => match K::get(self.data.column(1), base).partial_cmp(k) {
-                    Some(Less) => base + 1,
-                    _ => base,
-                },
-                None => base,
-            },
-            Greater => base,
+        // SAFETY: base is always in [0, size) because base <= mid.
+        let prefix = prefix_array.value(base);
+        let key = K::get(self.data.column(1), base);
+        let cmp = f((prefix, key));
+        if cmp == Ordering::Equal {
+            Ok(base)
+        } else {
+            let result = base + (cmp == Ordering::Less) as usize;
+            Err(result)
+        }
+    }
+
+    /// Returns the largest index where `prefixes[index] == prefix` or None if the provided prefix does not exist in the block.
+    #[inline]
+    fn find_largest_index_of_prefix<'me, K: ArrowReadableKey<'me>>(
+        &'me self,
+        prefix: &str,
+    ) -> Option<usize> {
+        // By design, will never find an exact match (comparator never evaluates to Equal). This finds the index of the first element that is greater than the prefix. If no element is greater, it returns the length of the array.
+        let result = self
+            .binary_search_by::<K, _>(|(p, _)| match p.cmp(prefix) {
+                Ordering::Less => Ordering::Less,
+                Ordering::Equal => Ordering::Less,
+                Ordering::Greater => Ordering::Greater,
+            })
+            .expect_err("Never returns Ok because the comparator never evaluates to Equal.");
+
+        if result == 0 {
+            // The first element is greater than the target prefix, so the target prefix does not exist in the block.
+            return None;
+        }
+
+        let prefix_array = self
+            .data
+            .column(0)
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .unwrap();
+
+        // `result` is the first index where the prefix is larger than the input (or the length of the array) so we want one element before this.
+        match prefix_array.value(result - 1).cmp(prefix) {
+            // We're at the end of the array, so the prefix does not exist in the block (all values are less than the prefix)
+            Ordering::Less => None,
+            // The prefix exists
+            Ordering::Equal => Some(result - 1),
+            // This is impossible
+            Ordering::Greater => None,
+        }
+    }
+
+    /// Finds the partition point of the prefix and key.
+    /// Returns the index of the first element that matches the target prefix and key. If no element matches, returns the index at which the target prefix and key could be inserted to maintain sorted order.
+    #[inline]
+    fn get_key_prefix_partition_point<'me, K: ArrowReadableKey<'me>>(
+        &'me self,
+        prefix: &str,
+        key: Option<&K>,
+    ) -> usize {
```

Comment:
I'm thinking if we should decompose this into two functions for better code clarity:
- `fn binary_search_prefix_key<K>(&self, prefix: &str, key: &K) -> Result<usize, usize>` where `Ok(i)` means the `(prefix, key)` combination is found at index `i` and `Err(i)` means it is not found but can be inserted at `i` to maintain order
- `fn find_smallest_index_with_prefix(&self, prefix: &str) -> Result<usize, usize>`, where `Ok(i)` means the prefix exists and starts at `i` and `Err(i)` means it is not found but can be inserted at `i` with a key value to maintain order

Reply:
> fn find_smallest_index_with_prefix(&self, prefix: &str) -> Result<usize, usize>, where Ok(i) means the prefix exists and starts at i and Err(i) means it is not found but can be inserted at i with a key value to maintain order

I think this has to return an Option, you can't infer an insert location from only a prefix

Reply:
I assume in the case the where the prefix does not exist, you can insert the prefix with any key value at that location without disturbing the order

Reply:
But an `Option` should be sufficient too

Reply:
🤦 yes you're right
I'll leave as an Option for now so find_smallest and find_largest are similar